packages/framework/tests/Unit/Component/Money/MoneyTest.php

```diff
@@ -0,0 +1,582 @@
+
+        Money::fromFloat($float, $scale);
+    }
+
+    /**
+     * @return \Iterator
+     */
+    public function invalidValuesFromFloatProvider(): Iterator
+    {
+        yield [NAN, 0];
+        yield [INF, 0];
+        yield [-INF, 0];
+        yield [NAN, 1];
+        yield [INF, 1];
+        yield [-INF, 1];
+    }
+
+    /**
+     * @dataProvider fromFloatProvider
+     * @param float $float
+     * @param int $scale
+     * @param string $expected
+     */
+    public function testFromFloat(float $float, int $scale, string $expected): void
+    {
+        $money = Money::fromFloat($float, $scale);
+
+        $this->assertEquals($expected, $money->toString());
+    }
+
+    /**
+     * @return \Iterator
+     */
+    public function fromFloatProvider(): Iterator
+    {
+        yield [0, 0, '0'];
+        yield [0, 1, '0.0'];
+        yield [0, 10, '0.0000000000'];
+        yield [1, 0, '1'];
+        yield [-1, 0, '-1'];
+        yield [10, 0, '10'];
+        yield [0.05, 1, '0.1'];
+        yield [0.5, 0, '1'];
+        yield [0.0001, 3, '0.000'];
+        yield [1.0001, 3, '1.000'];
+    }
+
+    public function testAddIsImmutable(): void
+    {
+        $money = Money::fromString('1');
+
+        $money->add(Money::fromString('1'));
+
+        $this->assertEquals('1', $money->toString());
+    }
+
+    /**
+     * @dataProvider addProvider
+     * @param string $a
+     * @param string $b
+     * @param string $expectedResult
+     */
+    public function testAdd(string $a, string $b, string $expectedResult): void
+    {
+        $moneyA = Money::fromString($a);
+        $moneyB = Money::fromString($b);
+
+        $result = $moneyA->add($moneyB);
+
+        $this->assertEquals($expectedResult, $result->toString());
+    }
+
+    /**
+     * @return \Iterator
+     */
+    public function addProvider(): Iterator
+    {
+        yield ['1', '1', '2'];
+        yield ['12.15', '34.965', '47.115'];
+        yield ['10', '-2', '8'];
+        yield ['1', '0.01', '1.01'];
+        yield ['0.5', '0.5', '1.0'];
+        yield ['1.525', '0.475', '2.000'];
+        yield ['1.00', '1.000', '2.000'];
+    }
+
+    public function testSubtractIsImmutable(): void
+    {
+        $money = Money::fromString('1');
+
+        $money->subtract(Money::fromString('1'));
+
+        $this->assertEquals('1', $money->toString());
+    }
+
+    /**
+     * @dataProvider subtractProvider
+     * @param string $a
+     * @param string $b
+     * @param string $expectedResult
+     */
+    public function testSubtract(string $a, string $b, string $expectedResult): void
+    {
+        $moneyA = Money::fromString($a);
+        $moneyB = Money::fromString($b);
+
+        $result = $moneyA->subtract($moneyB);
+
+        $this->assertEquals($expectedResult, $result->toString());
+    }
+
+    /**
+     * @return \Iterator
+     */
+    public function subtractProvider(): Iterator
+    {
+        yield ['2', '1', '1'];
+        yield ['12.15', '34.965', '-22.815'];
```

Comment:
please add also test case `higher presision, lower precision, higher precision`

Reply:
We have added one test case:  `yield ['1.000', '1.00', '0.000'];`